### ISSUE TITLE
Open source your code

### ISSUE THREAD

**Issue by tuxxy** [repotraffic/repotraffic-feedback#6]
I would use this service, but it's not open source.

You should give it a FOSS license. :)

**Comment by burnash** [comment]
I really want to open source it. However, making it open requires a bit of work. If there are enough people interested in helping to get this off the ground I'm totally open to this.

**Comment by petrsvihlik** [comment]
It would be very much appreciated. Especially for the sake of transparency. At the moment, I'm hesitant to give the repotraffic permissions to my repos.
![image](https://user-images.githubusercontent.com/9810625/44486013-b93b4b00-a652-11e8-8b6a-30f2172a12e5.png)
It would be great to see why certain permissions are needed and how they're used.

**Comment by george-haddad** [comment]
> I really want to open source it. However, making it open requires a bit of work. If there are enough people interested in helping to get this off the ground I'm totally open to this.

What kind of help are you looking for?

**Comment by burnash** [comment]
@george-haddad 
- Tests
- Cleaning code / refactoring
- Docs, installation guide, etc.

**Comment by burnash** [comment]
@petrsvihlik 
I'm not sure if open-sourcing actually increases the transparency for a hosted service. Unless everyone has access to the server that is running the _actuall_ software the transparency is limited.

-- BUT --

You've raised a very very important question that keeps popping up all the time. Sorry for the pause in response. The various distractions of work and life meant I wasn't able to respond in a timely manner.

*tl;dr* there's no way known to me to get the traffic data without the `public_repo` scope in an OAuth app. GitHub asks for read/write permission to access traffic data.

Repotraffic is a [GitHub OAuth](https://developer.github.com/apps/building-oauth-apps/understanding-scopes-for-oauth-apps/) app. 
The only scope that Repotraffic asks GitHub for is `public_repo`. Here's how it looks like in the app's Django settings responsible for configuring [Python Social Auth](https://python-social-auth.readthedocs.io/en/latest/):

```python
SOCIAL_AUTH_GITHUB_SCOPE = ['public_repo']
```

And this is the scope that you see on your screenshot. 

[GitHub API requires push access](https://developer.github.com/v3/repos/traffic/#views) to fetch traffic data:

> For repositories that you have push access to, the traffic API provides access to the information provided in the graphs section.

And responds with `403 Must have push access to repository` if the scope is missing.

This is an irritating issue and has been brought many times. Here's a discussion on [dear-github](https://github.com/dear-github/dear-github/issues/113) along with the list of other affected apps. As the OP, I have no interest in webhooks, wikis, keys, settings and et cetera. But for OAuth apps it's the only way to go.

There might be light at the end of the tunnel, though. GitHub has recently introduced a new type of 3rd party apps: [GitHub Apps](https://developer.github.com/apps/). It looks like these apps have a way to set a more granular permissions. 

Unfortunately, I haven't had enough time to do a thorough research to see if you can actually get traffic data without write permission using GitHub Apps. I wasn't able to find the answer in [GitHub Apps docs](https://developer.github.com/apps/about-apps/). If someone is interested in doing the research and contributing, please let me know.

**Comment by fedorov** [comment]
>> What kind of help are you looking for?
> * Tests.
> * Cleaning code / refactoring.
> * Docs, installation guide, etc.

This is exactly the kind of help you have a lot better chances to get AFTER you open source your code. If your users find the source code useful, they will feel motivated to help maintain it. I think you will need a lot of luck to motivate someone to do refactoring of a code one has never seen.

**Comment by fedorov** [comment]
cc: @jcfr re https://github.com/isaacs/github/issues/399#issuecomment-285505598

**Comment by fedorov** [comment]
Sorry about that. I thought you were looking for help with the items you mentioned from other people.

**Comment by burnash** [comment]
Are you interested in contributing to the project?

**Comment by fedorov** [comment]
If the project was open source, and I could try it out, I would have a chance to see if there is an intersection between my needs, my skills, and my available time, and decide if it makes sense for me to contribute, and in what capacity.

**Comment by burnash** [comment]
How much time/capacity do you think you could contribute to the project?

**Comment by george-haddad** [comment]
I was also thinking that should it be opensource then anyone can just fork it and submit a PR at any moment they have time and capacity to do so. Then again, I would love to contribute but not sure what the code base is written in.

**Comment by burnash** [comment]
@george-haddad Nothing fancy, it's a Python 2.7 project that uses Django 1.x. and python-social-auth. Here're some stats from [cloc](https://github.com/AlDanial/cloc):

```
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
Python                          17            212             57            647
HTML                             7             25             13            257
CSS                              2             11              0             48
Javascript                       5              6              2             44
make                             1              5              0             12
-------------------------------------------------------------------------------
SUM:                            32            259             72           1008
-------------------------------------------------------------------------------
```

@fedorov @george-haddad let me know if this matches your skills/capacity and you're available to help to open source the project. If you are, I'll send you an invite to the private repo.

**Comment by burnash** [comment]
@fedorov sorry, I don't understand the point of making your comments then.

**Comment by george-haddad** [comment]
> @george-haddad Nothing fancy, it's a Python 2.7 project that uses Django 1.x. and python-social-auth. Here're some stats from [cloc](https://github.com/AlDanial/cloc):
> 
> ```
> -------------------------------------------------------------------------------
> Language                     files          blank        comment           code
> -------------------------------------------------------------------------------
> Python                          17            212             57            647
> HTML                             7             25             13            257
> CSS                              2             11              0             48
> Javascript                       5              6              2             44
> make                             1              5              0             12
> -------------------------------------------------------------------------------
> SUM:                            32            259             72           1008
> -------------------------------------------------------------------------------
> ```
> @fedorov @george-haddad let me know if this matches your skills/capacity and you're available to help to open source the project. If you are, I'll send you an invite to the private repo.

Right off the bat, I'm not a python dev, I do mostly Java, nodejs and reactjs. I would say that I can't help you because of that, though if I had access to the source code then I might find something that I can do (depending on the gh issues).

It's very tough to give an answer, as things stand right now Python devs are targeted, but a decent dev can figure out python and implement a feature depending if the code is readable and well structured and if the feature isn't so complex. So in this respect being opensource project would attract many eyes and talent, right now all I know is that 647 lines of code are in python and 44 lines of code are in JS and I know more JS than Python, I could do some Python but no insight into the complexity of the code-base to see if it's even worth it to put my available time and effort into it.

**Comment by burnash** [comment]
@george-haddad Thank you for your consideration and for offering help. I really appreciate it.

I gave this discussion another thought.

So far I'm not convinced that I and the project by that extent will benefit in any significant way from open-sourcing it.

What the project needs is quality maintenance (not many new features). What the project does not need are endless discussions and unsolicited advice. For now, I'm not convinced that open-sourcing the project would result in maintenance improvements. However, I did get a hint of possible downsides in this very discussion.

I believe there are some other ways to address maintenance quality. I'll share more details once I explore them better.

For the time being the source code of the project will remain unpublished. If anyone is willing to explore open source alternative to Repotraffic, you may take a look at https://github.com/M165437/gh-traffic

Thanks again to all participants, I'm closing this issue.

**Comment by tuxxy** [comment]
The reason I opened this issue was because I didn't like your code
requesting all the permissions it did.

Thanks for the thoughts anyway, we'll continue to not use it.

**Comment by burnash** [comment]
@tuxxy I think I explained the permissions situation in details [above](https://github.com/repotraffic/repotraffic-feedback/issues/6#issuecomment-432638356). I also made a point that open-sourcing something per se does not increase transparency for a hosted service.

Some people will continue to use the project, others don't. And it's OK. Please don't make a drama out of it.

**Comment by burnash** [comment]
PS: I'm sorry, I don't have enough time to discuss this further.
